### PR TITLE
GUI: allow sorting Your Orders table by different colums

### LINKED ISSUE #878: allow sorting Your Orders table by different colums e.g. rate

### PR DESCRIPTION
**Pull request by amassarwi**
This adds sort feature to `Your Orders` table in dexc's GUI.

By default orders will be sorted descending by age.
Users could change to sort by another header by clicking on it, or change the sort direction by clicking
on the currently chosen sort column.

Closes #878.

### CODE REVIEW

File: client/webserver/site/src/js/markets.js [hunk 1]
```diff
@@ -188,14 +192,61 @@ export default class MarketsPage extends BasePage {
...
+
+    const unsetOrdersSortColClasses = () => {
+      const camelCaseKey = strCamelCase(this.ordersSortKey)
+      page[`live${camelCaseKey}Col`].classList.remove(`sorted-${this.ordersSortDirection}`)
+    }
+
+    const setOrdersSortColClasses = () => {
+      const camelCaseKey = strCamelCase(this.ordersSortKey)
+      page[`live${camelCaseKey}Col`].classList.add(`sorted-${this.ordersSortDirection}`)
+    }
```
Comment (buck54321) [inline client/webserver/site/src/js/markets.js:246]: This type of string substitution with case change for object keys is messy. We need to find a better solution. 

My suggestion is to get rid of the IDs altogether on the `th` elements. Use `data-` attributes instead. 

```html
<th data-ordercol="liveTypeCol">
    <div>Type</div>
    <div class="sort-arrow" />
</th>
<th data-ordercol="liveSideCol">
    <div>Side</div>
    <div class="sort-arrow" />
</th>
...
```

This would allow you to do things like.

```js
page.liveTable.querySelectorAll('[data-ordercol]').forEach(th => th.classList.remove('sorted-asc', 'sorted-dsc'))
```

and set it like

```js
page.liveTable.querySelector(`[data-ordercol=${key}]`).classList.add(cls))
```

File: client/webserver/site/src/css/market_dark.scss [hunk 2]
```diff
@@ -1,5 +1,5 @@
 body.dark {
-  input[type=number] {
+  input[type='number'] {
```
Comment (buck54321) [inline client/webserver/site/src/css/market_dark.scss:2]: Can you justify this change? From all appearances, the new syntax appears to be incorrect.

File: client/webserver/site/src/css/market_dark.scss [hunk 3]
```diff
@@ -133,4 +133,12 @@ body.dark {
...
     }
   }
+
+  #liveTable {
+    th {
+      .sort-arrow {
+        background-image: url("/img/arrow_down_dark.svg");
+      }
+    }
+  }
```
Comment (buck54321) [inline client/webserver/site/src/css/market_dark.scss:143]: Look into how we handle icons in **icomoon.svg** and **icons.scss**.

File: client/webserver/site/src/js/markets.js (hunk 1 above)
Comment (buck54321) [inline client/webserver/site/src/js/markets.js:246]: Please check the sense of direction on these arrows. I'm seeing

image

To me, the up arrow usually indicates ascending, such that the values will rise as you move down the table. I've checked a couple of other tables on other sites and my assumption appears to jive.

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/webserver/site/src/img/arrow_down.svg [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/webserver/site/src/img/arrow_down.svg:1]: It's bee a while, let me see if I can remember...

File: client/webserver/site/src/css/market_dark.scss (hunk 2 above)
Comment (amassarwi) [reply client/webserver/site/src/css/market_dark.scss:2]: it's my autosave prettier formatting, will revert.

File: client/webserver/site/src/js/markets.js (hunk 1 above)
Comment (amassarwi) [reply client/webserver/site/src/js/markets.js:246]: 👍

File: client/webserver/site/src/js/markets.js [hunk 5]
```diff
@@ -779,13 +830,47 @@ export default class MarketsPage extends BasePage {
...
+          (a.rate - b.rate)
+      case 'qty':
+        return (a, b) => (this.ordersSortDirection === 'dsc' ? -1 : 1) *
+          (a.qty - b.qty)
+      case 'filled':
+        return (a, b) => (this.ordersSortDirection === 'dsc' ? -1 : 1) *
+          (a.qty - b.qty)
+      case 'settled':
+        return (a, b) => (this.ordersSortDirection === 'dsc' ? -1 : 1) *
+          (a.qty - b.qty)
```
[the remark this answers is not in the document]
Comment (amassarwi) [reply client/webserver/site/src/js/markets.js:856]: `a.settled - b.settled` might be enough.

File: client/webserver/site/src/js/markets.js [hunk 6]
```diff
@@ -779,13 +830,47 @@ export default class MarketsPage extends BasePage {
...
+  ordersSortCompare () {
+    switch (this.ordersSortKey) {
+      case 'age':
+        return (a, b) => (this.ordersSortDirection === 'dsc' ? 1 : -1) *
+          (a.epoch - b.epoch)
+      case 'type':
+        return (a, b) => (this.ordersSortDirection === 'dsc' ? -1 : 1) *
+          (a.type - b.type)
+      case 'rate':
+        return (a, b) => (this.ordersSortDirection === 'dsc' ? -1 : 1) *
```
[the remark this answers is not in the document]
Comment (amassarwi) [reply client/webserver/site/src/js/markets.js:846]: right, good idea!

**Comment by amassarwi** [comment]
@buck54321 Updated.

File: client/webserver/site/src/js/markets.js [hunk 7]
```diff
@@ -779,13 +828,42 @@ export default class MarketsPage extends BasePage {
     }
   }
 
+  /*
+   * ordersSortCompare returns sort compare function according to the active
+   * sort key.
```
Comment (amassarwi) [inline client/webserver/site/src/js/markets.js:833]:
```suggestion
   * sort key and direction.
```

File: client/webserver/site/src/img/arrow_down.svg (hunk 4 above)
[the remark this answers is not in the document]
Comment (chappjc) [reply client/webserver/site/src/img/arrow_down.svg]: Which way did you go on this process @amassarwi?  We should document the possible approaches.

[the remark this answers is not in the document]
Comment (amassarwi) [reply client/webserver/site/src/img/arrow_down.svg]: icomoon, was pretty straightforward

**Review by JoeGruffins — COMMENTED** [review]
Works great! 

Except it looks like the "Settled" column does not sort like the others, and "Filled" and "Status" don't always sort correctly. (see pics)

Also, I feel like the arrow is too big. My opinion.

filledsort
statussort

**Comment by chappjc** [comment]
Working very well overall for me, but I noticed the same things that @JoeGruffins pointed out.  Also that the column alignment seems to jump around strangely when changing columns to sort by.  Agree that the arrow is a touch too large.  Here's a video of some of the quirks I observed:  

https://user-images.githubusercontent.com/9373513/107263568-85f9fb80-6a07-11eb-9b76-8d4767482825.mp4

**Comment by amassarwi** [comment]
@JoeGruffins @chappjc Updated mates, thank you.

**Review by JoeGruffins — COMMENTED** [review]
Filled and Settled still look off sometimes.
<details closed>
<summary>pics</summary>

filledsort
settledsort
<details>

**Comment by amassarwi** [comment]
@JoeGruffins Hopefully fixed for good now.

**Review by JoeGruffins — APPROVED** [review]

File: client/webserver/site/src/css/market.scss [hunk 8]
```diff
@@ -420,6 +420,47 @@ table.balance-table button:hover {
   td {
     padding: 2px 10px;
     white-space: nowrap;
+    flex: 3;
+
+    &.last {
+      flex: 1;
```
Comment (buck54321) [inline client/webserver/site/src/css/market.scss:426]: It wasn't clear why you were using flex here now, so I went through and removed the flex and it still worked fine. Was there another reason to use flex that I'm not seeing?

https://github.com/decred/dcrdex/commit/beb9ed2496449163b48abcfb63602536fa46c230

The one thing I did notice is that the arrow icon seems to be off center, so I loaded the svg into icomoon and got this

image
which seems like a big clue that something was wrong with your icon. I would start by using an svg editor to make a square icon with the arrow centered vertically.

File: client/webserver/site/src/js/markets.js [hunk 9]
```diff
@@ -188,14 +191,60 @@ export default class MarketsPage extends BasePage {
     ws.registerRoute(epochOrderRoute, data => { this.handleEpochOrderRoute(data) })
     // Bind the wallet unlock form.
     bindOpenWallet(app, page.openForm, async () => { this.openFunc() })
-    // Create a wallet
+    // Create a wallet.
```
Comment (buck54321) [inline client/webserver/site/src/js/markets.js:194]: This is not a sentence so doesn't get a full stop, nor is it relevant to this PR.

File: client/webserver/site/src/js/markets.js [hunk 10]
```diff
@@ -779,13 +828,47 @@ export default class MarketsPage extends BasePage {
...
+
   /* refreshActiveOrders refreshes the user's active order list. */
   refreshActiveOrders () {
     const page = this.page
     const metaOrders = this.metaOrders
     const market = this.market
     for (const oid in metaOrders) delete metaOrders[oid]
-    const orders = app.orders(market.dex.host, marketID(market.baseCfg.symbol, market.quoteCfg.symbol))
+    const orders = app.orders(market.dex.host, marketID(market.baseCfg.symbol,
+      market.quoteCfg.symbol))
```
Comment (buck54321) [inline client/webserver/site/src/js/markets.js:868]: No need to indent. Please undo.

File: client/webserver/site/src/js/markets.js [hunk 11]
```diff
@@ -799,7 +882,9 @@ export default class MarketsPage extends BasePage {
         this.setMarkers()
       })
       updateUserOrderRow(row, ord)
-      if (ord.type === Order.Limit && (ord.tif === Order.StandingTiF && ord.status < Order.StatusExecuted)) {
+      if (ord.type === Order.Limit && (ord.tif === Order.StandingTiF &&
+        ord.status < Order.StatusExecuted)
```
Comment (buck54321) [inline client/webserver/site/src/js/markets.js:886]: No need to indent. Please undo.

File: client/webserver/site/src/js/order.js [hunk 12]
```diff
@@ -30,9 +30,8 @@ export default class OrderPage extends BasePage {
     ])
 
     if (page.cancelBttn) {
-      Doc.bind(page.cancelBttn, 'click', () => {
-        this.showForm(page.cancelForm)
-      })
+      Doc.bind(page.cancelBttn, 'click', () =>
+        this.showForm(page.cancelForm))
```
Comment (buck54321) [inline client/webserver/site/src/js/order.js:34]: Please undo the re-formatting in this file. If you really feel it's important to reformat this file, submit a separate PR please. But in all honesty, nothing here needs reformatting.

File: client/webserver/site/src/js/markets.js [hunk 13]
```diff
@@ -188,14 +191,60 @@ export default class MarketsPage extends BasePage {
...
+      let direction
+      switch (this.ordersSortDirection) {
+        case 1:
+          direction = 'asc'
+          break
+        case -1:
+          direction = 'dsc'
+          break
+      }
+      return `sorted-${direction}`
```
Comment (buck54321) [inline client/webserver/site/src/js/markets.js:232]:
```
if (this.ordersSortDirection === 1) return 'sorted-asc'
return 'sorted-dsc'
```

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/webserver/site/src/js/markets.js (hunk 13 above)
Comment (amassarwi) [reply client/webserver/site/src/js/markets.js:232]: 👑

**Comment by chappjc** [comment]
Can you squash and rebase?  Given the harness changes churning in this PR, I'd actually do it in two separate steps. That is, first squash without actually changing the branching commit `git rebase -i $(git merge-base master HEAD)` then rebase.

**Comment by chappjc** [comment]
I'm not sure if this is because of this PRs changes, but in testing the latest commits, I have a 50% filled order booked that is also settling but it only says booked and not booked/settling or whatever it usually says (it usually indicates settling I think, right?)

image

image

**Comment by amassarwi** [comment]
@chappjc squashed & rebased.
regarding the booked/settling I'm pretty sure it's not related to this PR.

**Comment by chappjc** [comment]
Yeah, I'll have to check on master again.  Everything looks good though now.  All sorts working well and colum.n alignments look good.

**Review by chappjc — COMMENTED** [review]
All sorts working well and column alignments look good.

**Comment by chappjc** [comment]
Looks like @amassarwi hit your review items, @buck54321.
